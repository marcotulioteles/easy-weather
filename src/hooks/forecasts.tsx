import { useContext, useEffect, createContext, ReactNode, useState, SetStateAction } from 'react';
import { oneCallWeatherDataAPI, directGeocodingAPI } from '../services/api';

type ForecastsProviderProps = {
  children: ReactNode;
}

type HourlyForecastData = {
  dt: number;
  temp: number;
  feels_like: number;
  humidity: number;
  clouds: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string;
    }
  ]
}

type DailyForecastData = {
  dt: number;
  temp: {
    day: number;
  };
  feels_like: {
    day: number;
  };
  humidity: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string
    }
  ];
  clouds: number;
}

type CurrentForecastData = {
  temp: number;
  feels_like: number;
  humidity: number;
  clouds: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string
    }
  ];
}

interface ForecastData {
  current: CurrentForecastData;
  hourly: HourlyForecastData[];
  daily: DailyForecastData[];
}

interface LocationDataResponse {
  name: string;
  country: string;
  state: string;
}

interface ForecastContextProps {
  locationInput: string;
  locationResponse: LocationDataResponse;
  setLocationInput: React.Dispatch<SetStateAction<string>>;
  error: boolean;
  forecast: ForecastData;
  loading: boolean;
}

export const ForecastsContext = createContext({} as ForecastContextProps);

function ForecastsProvider({ children }: ForecastsProviderProps) {
  const [locationInput, setLocationInput] = useState('');
  const [locationResponse, setLocationResponse] = useState<LocationDataResponse>({} as LocationDataResponse);
  const [forecast, setForecast] = useState<ForecastData>({} as ForecastData);
  const [error, setError] = useState(false);
  const [loading, setLoading] = useState(false);

  async function fetchForecast(locationInput: string) {
    try {
      setLoading(true);

      const response = await directGeocodingAPI
        .get(`direct?q=${locationInput}&appid=${process.env.NEXT_PUBLIC_APPID}`);
      const locationData = response.data;

      const locationFormatted = {
        name: locationData[0].name,
        country: locationData[0].country,
        state: locationData[0].state
      }

      setLocationResponse(locationFormatted);

      if (locationData) {
        const response = await oneCallWeatherDataAPI
          .get(`onecall?lat=${locationData[0].lat}&lon=${locationData[0].lon}&exclude=minutely,alerts&appid=${process.env.NEXT_PUBLIC_APPID}`)
        const forecastData = response.data;

        const dataFormatted = {
          current: {
            temp: forecastData.current.temp,
            feels_like: forecastData.current.feels_like,
            humidity: forecastData.current.humidity,
            clouds: forecastData.current.clouds,
            wind_speed: forecastData.current.wind_speed,
            weather: forecastData.current.weather,
          },
          hourly: forecastData.hourly.map((hour: HourlyForecastData) => {
            return {
              dt: hour.dt,
              temp: hour.temp,
              feels_like: hour.feels_like,
              humidity: hour.humidity,
              clouds: hour.clouds,
              wind_speed: hour.wind_speed,
              weather: hour.weather.map(weather => {
                return {
                  description: weather.description,
                  icon: weather.icon
                }
              })
            }
          }),
          daily: forecastData.daily.map((day: DailyForecastData) => {
            return {
              dt: day.dt,
              temp: day.temp,
              feels_like: day.feels_like,
              humidity: day.humidity,
              clouds: day.clouds,
              wind_speed: day.wind_speed,
              weather: day.weather.map(weather => {
                return {
                  description: weather.description,
                  icon: weather.icon
                }
              })
            }
          })
        }

        setForecast(dataFormatted);
        setLoading(false);
      }
    } catch (error) {
      setError(true);
    }
  }

  useEffect(() => {
    if (locationInput) fetchForecast(locationInput);
  }, [locationInput]);

  return (
    <ForecastsContext.Provider value={
      {
        locationInput,
        locationResponse,
        setLocationInput,
        error,
        forecast,
        loading
      }}>
      {children}
    </ForecastsContext.Provider>
  );
};

function useGetForecasts() {
  const context = useContext(ForecastsContext);

  return context;
}

export { ForecastsProvider, useGetForecasts };