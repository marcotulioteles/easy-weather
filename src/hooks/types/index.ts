export type HourlyForecastData = {
  dt: number;
  temp: number;
  feels_like: number;
  humidity: number;
  clouds: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string;
    }
  ]
}

export type DailyForecastData = {
  dt: number;
  temp: {
    min: number;
    max: number;
  };
  feels_like: {
    day: number;
  };
  humidity: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string
    }
  ];
  clouds: number;
}

export type CurrentForecastData = {
  temp: number;
  min_temp: number;
  max_temp: number;
  feels_like: number;
  humidity: number;
  clouds: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string
    }
  ];
}

export interface ForecastData {
  current: CurrentForecastData;
  hourly: HourlyForecastData[];
  daily: DailyForecastData[];
}

export interface LocationData {
  name: string;
  country: string;
  state: string;
}