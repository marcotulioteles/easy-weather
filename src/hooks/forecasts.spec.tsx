import { fireEvent, render } from '@testing-library/react';
import React, { FormEvent, useState } from 'react';
import { ForecastsContext, ForecastsProvider, useGetForecasts } from './forecasts';
import { ForecastData, LocationData } from './types';

const Component = () => {
  const {
    locationInput,
    locationResponse,
    setLocationInput,
    error,
    forecast,
    loading,
    isEmpty
  } = useGetForecasts();

  const [locationInputValue, setLocationInputValue] = useState('');

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setLocationInput(locationInputValue)
  }

  return (
    <>
      <form onSubmit={handleSubmit}>
        <input
          type='text'
          onChange={event => setLocationInputValue(event.target.value)}
          data-testid='input-location'
        />
        <button type='submit' data-testid='button-submit'>Click here</button>
      </form>
      <p data-testid='location-name'>{locationInput}</p>
      <p data-testid='location-country'>{locationResponse.country}</p>
      {error && <p>This is a error!</p>}
      {loading && <p>This a loading...</p>}
      {isEmpty && <p>This is a empty state!</p>}
      {forecast?.current && <h1>{forecast.current.temp}</h1>}
    </>
  );
}

describe('Forecast Context', () => {
  it('should call de setLocationInput function', () => {
    const component = render(
      <ForecastsProvider>
        <Component />
      </ForecastsProvider>
    );

    expect(component).toMatchSnapshot();
  });

  it('should render the location input inside the document', () => {
    const { getByTestId } = render(
      <ForecastsProvider>
        <Component />
      </ForecastsProvider>
    );

    const input = getByTestId('input-location');
    const button = getByTestId('button-submit');

    fireEvent.change(input, { target: { value: 'any city' } });
    fireEvent.click(button);

    expect(getByTestId('location-name')).toHaveTextContent('any city');
  });
});