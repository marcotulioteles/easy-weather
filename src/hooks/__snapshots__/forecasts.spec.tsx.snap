// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forecast Context should call de setLocationInput function 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <input
          data-testid="input-location"
          type="text"
        />
        <button
          data-testid="button-submit"
          type="submit"
        >
          Click here
        </button>
      </form>
      <p
        data-testid="location-name"
      />
      <p
        data-testid="location-country"
      />
    </div>
  </body>,
  "container": <div>
    <form>
      <input
        data-testid="input-location"
        type="text"
      />
      <button
        data-testid="button-submit"
        type="submit"
      >
        Click here
      </button>
    </form>
    <p
      data-testid="location-name"
    />
    <p
      data-testid="location-country"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
